core/src/apps/cardano/helpers/staking_use_cases.py

```diff
@@ -0,0 +1,57 @@
+from trezor.messages import CardanoAddressType
+
+from ..address import get_public_key_hash
+
+if False:
+    from typing import List
+    from trezor.messages import CardanoAddressParametersType
+    from . import seed
+
+
+"""
+Used as a helper when deciding what warnings we should
+display to the user during get_address and sign_tx depending
+on the type of address and its parameters.
+"""
+
+
+NO_STAKING = 0
+SAME_ACCOUNT = 1
+SAME_HASH = 2
+DIFFERENT_ACCOUNT = 3
+DIFFERENT_HASH = 4
+POINTER_ADDRESS = 5
+REWARD_ADDRESS = 6
+
+
+def get(
+    keychain: seed.Keychain, address_parameters: CardanoAddressParametersType
+) -> int:
+    address_type = address_parameters.address_type
+    if address_type == CardanoAddressType.BASE:
+        spending_account_staking_path = _path_to_staking_path(
+            address_parameters.spending_key_path
+        )
+        if address_parameters.staking_key_path:
+            if address_parameters.staking_key_path != spending_account_staking_path:
+                return DIFFERENT_ACCOUNT
+            else:
+                return SAME_ACCOUNT
+        else:
+            staking_key_hash = get_public_key_hash(
+                keychain, spending_account_staking_path
+            )
+            if address_parameters.staking_key_hash != staking_key_hash:
+                return DIFFERENT_HASH
+            else:
+                return SAME_HASH
+    elif address_type == CardanoAddressType.POINTER:
+        return POINTER_ADDRESS
+    elif address_type == CardanoAddressType.REWARD:
+        return REWARD_ADDRESS
+    else:
+        return NO_STAKING
+
+
+def _path_to_staking_path(path: List[int]) -> List[int]:
```

Comment:
this seems to silently assume that `path` has at least 3 components; weird things will happen if `path == []` for example

Reply:
True, I've added a check for that [here](https://github.com/trezor/trezor-firmware/pull/1112/commits/ab42be29b0dca46baf1d817a81e63856df320965).

Reply:
this will fail without explanation when you set, e.g., `address_n=[]`

Re-reading this, I'm not sure what exactly the code accomplishes.

Every address in the format `m/shelley'/cardano'/account'/*` must have staking path `m/shelley'/cardano'/account'/2/0`, otherwise it is `DIFFERENT_ACCOUNT` ... but that is ALSO true for `m/shelley'/cardano'/account'/0/0` or `/2/1` or any other variation that is in fact within the same account.

(for that matter: is the staking path always `/2/0`? why not `/2/address_index` ?)

Also if someone used a custom path scheme, e.g. something like Electrum's Purpose48, where there are four hardened components, this would again return incorrect results.

---

i'd propose approaching this a little differently:

* Fold "same account" and "same hash" into a single case, something like `STAKING_MATCH`
  This case has strong checks: `address_n` must conform to `validate_full_path`, and the path must match SHELLEY purpose - in addition to what's already there, i.e., (a) `address_n_staking` is what `_path_to_staking_path()` says, or (b) provided key hash matches that
* ANY OTHER CASE (for base address) is `STAKING_MISMATCH`; for convenience you can differentiate between "account" and "key", but ISTM the caller is a better place for that. For instance in `get_address._show_staking_warning` you don't need two branches, you can call `show_warning_address...()` with all params supplied, confident that at some point `_validate_base_address_staking_info()` must have happened.
* (aside: couldn't `REWARD` also fold into `STAKING_MATCH`?)
* reword the warnings. I think "does not belong to account" is the right direction, but "belong to" isn't strictly true. How about "does not match the account"?

this way as long as the user upholds the schema, they're OK.
when they uphold the schema but use unexpected staking info, we warn them
when they don't uphold the schema, we don't know anything so we warn them too

Reply:
>(for that matter: is the staking path always /2/0? why not /2/address_index ?)

We've been told by IOHK, that only 2/0 will be used. We've also been refered to this [document](https://github.com/input-output-hk/implementation-decisions/blob/e2d1bed5e617f0907bc5e12cf1c3f3302a4a7c42/text/1852-hd-chimeric.md#drawbacksimplications). 

What you propose sounds reasonable and I will rewrite it. However, this still won't work for the paths with four hardened components, right? (I'm not familiar with Electrum's Purpose48). What I'm asking is, whether it is OK to keep the current implementation of `_path_to_staking_path` or if that needs to be changed?

Reply:
In what i'm proposing, it is only supposed to work with the standard BIP44-like paths. So the fact that it returns "incorrect" result for a path with four hardened components doesn't matter: that path fails `validate_full_path` and so it is automatically a mismatch.

So yes, you can keep `_path_to_staking_path` as is.

Reply:
Ok, I've updated the code [here](https://github.com/trezor/trezor-firmware/pull/1112/commits/a68b518e623927c1b12a78c894b8a67ad3414f3c). I have two questions though.
1. When deriving a base address we check that either `address_n_staking` or `staking_key_hash` is set. And we also validate `address_n_staking` if it is set. However, we never check that only one of `address_n_staking` or `staking_key_hash` is set. Is that OK? Since we now pass both parameters to `show_warning_address...()`. We always check for `address_n_staking` first.
2. Warnings text - is that OK? Perhaps "Change address staking account does not match the current account" would be better? The current (updated) wording seems a bit cleaner to me although I am not sure whether "rights" can match anything.

Reply:
> 1. When deriving a base address we check that either `address_n_staking` or `staking_key_hash` is set. And we also validate `address_n_staking` if it is set. However, we never check that only one of `address_n_staking` or `staking_key_hash` is set. Is that OK? Since we now pass both parameters to `show_warning_address...()`. We always check for `address_n_staking` first.

Ah, I was under the impression that `_validate_base_address_staking_info` checks that too.
My bad. It doesn't matter too much, but let's fix it anyway. Please change the check as indicated in the followup review comment.

> 2. Warnings text - is that OK? Perhaps "Change address staking account does not match the current account" would be better? The current (updated) wording seems a bit cleaner to me although I am not sure whether "rights" can match anything.

This is OK by me - but feel free to change it further if you come up with something better.